### PR TITLE
[master] Add Dockerfiles and Upload Docker images CI [AP-1504]

### PR DESCRIPTION
**Pull request by y0urself**
**What**:

AP-1504

**Why**:

**How**:

**Checklist**:

- [ ] Tests
- [ ] [CHANGELOG](https://github.com/greenbone/gsa/blob/master/CHANGELOG.md) Entry
- [ ] Labels for ports to other branches

### CODE REVIEW

File: .github/workflows/ci-c.yml [hunk 1]
```diff
@@ -25,11 +25,37 @@ jobs:
...
+    container: ${{ matrix.container }}
+    env:
+      BUILD_TYPE: ${{ matrix.build_type }}
+    steps:
+      - uses: actions/checkout@v2.3.4
+      - name: Configure and compile
+        run: |
+          mkdir build && rm -rf .git && cd build/ \
+            && cmake -DCMAKE_BUILD_TYPE=$BUILD_TYPE .. && make install
+        working-directory: ${{ github.WORKSPACE }}
```
Comment (bjoernricks) [inline .github/workflows/ci-c.yml:61]:
```suggestion
        working-directory: ${{ github.WORKSPACE }}

```

File: .github/workflows/docker-ci.yml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      - name: Build and push
+        uses: docker/build-push-action@v2.5.0
+        with:
+          context: .
+          push: true
+          tags: ${{ github.repository }}-${{ env.VERSION }}-${{matrix.compiler}}-build
+          file: .docker/build/Dockerfile
+          build-args: |
+            VERSION=${{ env.VERSION }}
+            COMPILER=${{matrix.compiler}}
```
Comment (bjoernricks) [inline .github/workflows/docker-ci.yml:90]:
```suggestion
            COMPILER=${{matrix.compiler}}

```

File: .docker/testing/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+# We want gvm-libs to be ready so we use the build docker image of gvm-libs
+FROM greenbone/gvm-libs-$VERSION-$COMPILER-build
+
+# This will make apt-get install without question
+ARG DEBIAN_FRONTEND=noninteractive
+
+# Install curl and gnupg to add remote repositories
+RUN apt-get update && apt-get install --assume-yes \
+    curl \
+    gnupg \
```
Comment (bjoernricks) [inline .docker/testing/Dockerfile:17]:
```suggestion
    gnupg \
    yarnpkg \
    nodejs \
```

File: .docker/testing/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    | apt-key add - && \
+    echo "deb http://dl.yarnpkg.com/debian/ stable main" \
+    > /etc/apt/sources.list.d/yarn.list
+
+# Add repository for node.js
+RUN curl --silent --show-error \
+    https://deb.nodesource.com/gpgkey/nodesource.gpg.key \
+    | apt-key add - && \
+    echo "deb http://deb.nodesource.com/node_10.x buster main" \
+    > /etc/apt/sources.list.d/nodesource.list
```
Comment (bjoernricks) [inline .docker/testing/Dockerfile:32]: yarn and nodejs 10 are provided as debian packages already. No need to install them here.
```suggestion
```

File: .docker/testing/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+# Install node.js and yarn for gsa
+RUN apt-get update && apt-get install --assume-yes \
+    nodejs \
+    yarn \
+    && rm -rf /var/lib/apt/lists/*
```
Comment (bjoernricks) [inline .docker/testing/Dockerfile:38]:
```suggestion
```

File: .docker/testing/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+# Install Debian core dependencies required for building gsad
+RUN apt-get update && apt-get install --assume-yes \
+    libmicrohttpd-dev \
+    libxml2-dev \
+    libxslt1-dev \
```
Comment (bjoernricks) [inline .docker/testing/Dockerfile:44]: Should be installed above in the first apt-get install statement already. Every RUN statement in a Dockerfile creates an extra image layer.

**Review by bjoernricks — CHANGES_REQUESTED** [review]

File: .docker/testing/Dockerfile (hunk 5 above)
Comment (y0urself) [reply .docker/testing/Dockerfile:38]: Why do you think we do not need node.js? We use this image for our JavaScript testing ..., don't we?

File: .docker/testing/Dockerfile (hunk 4 above)
Comment (y0urself) [reply .docker/testing/Dockerfile:32]: The root image we use is debian-slim. Can you show me, where you did find the information, that node is provided here?

Comment (y0urself) [reply .docker/testing/Dockerfile:32]: I have looked into the https://github.com/debuerreotype/docker-debian-artifacts/tree/a4f413e5917d5917fb2343c0c37ea0728114c084/buster/slim tarball and couldn't find yarn or node or gnupg. :/

**Comment by y0urself** [comment]
AP-1504

**Review by bjoernricks — APPROVED** [review]

**Pull request merged**